include/unifex/task.hpp

```diff
@@ -222,54 +222,71 @@ struct _result_and_unhandled_exception final {
    * This is used to share the implementation of result handling between
    * type-specific promise types.
    */
-  struct type {
-    void unhandled_exception() noexcept {
-      expected_.reset_value();
-      unifex::activate_union_member(
-          expected_.exception_, std::current_exception());
-      expected_.state_ = _state::exception;
+  struct type { //parametize on noexcept, noexcept terminates in body
+    void unhandled_exception() noexcept { // will be invoked in the catch block of the coroutine
+      if constexpr (nothrow) {
+        std::terminate();
+      } else {
+        expected_.reset_value();
+        unifex::activate_union_member(
+            expected_.exception_, std::current_exception());
+        expected_.state_ = _state::exception;
+      }
+    }
+
+    decltype(auto) result() noexcept(nothrow) {
+      if constexpr (nothrow) {
+        return std::move(expected_).get();
+      } else {
+        if (expected_.state_ == _state::exception) {
+          std::rethrow_exception(std::move(expected_.exception_).get());
+        }
+        return std::move(expected_.value_).get();
+      }
     }
 
-    decltype(auto) result() {
-      if (expected_.state_ == _state::exception) {
-        std::rethrow_exception(std::move(expected_.exception_).get());
+    template <typename... Args>
+    // todo: consider if this should be nothrow or not
+    void set_value(Args&&... values) {
+      if constexpr (nothrow) {
+        expected_.construct(static_cast<Args&&>(values)...);
+      } else {
+        this->expected_.reset_value();
+        unifex::activate_union_member(
+            this->expected_.value_, static_cast<Args&&>(values)...);
+        this->expected_.state_ = _state::value;
       }
-      return std::move(expected_.value_).get();
     }
 
-    _expected<T> expected_;
+    std::conditional_t<nothrow, manual_lifetime<T>, _expected<T>> expected_;
```

Comment:
I think I'd rename this to `result_` now that it's not necessarily an `_expected<T>`.
```suggestion
    std::conditional_t<nothrow, manual_lifetime<T>, _expected<T>> result_;
```